File: GameEngine_IOS/GameEngine/Shaders/Entities/EntityShaderManager.m
#import "EntityShaderManager.h"

/**
 * Manager of the shader files that are going to be load to render the 3D
 */
@implementation EntityShaderManager
{
@private
    
    // Uniform index.
    enum
    {
        PROJECTION_MATRIX,
        VIEW_MATRIX,
        TRANSFORMATION_MATRIX,
        NORMAL_MATRIX,
        LIGHT_POSITION,
        LIGHT_COLOR,
        SHINE_DAMPER,
        REFLECTIVITY,
        SKY_COLOR,
        NORMALS_POINTING_UP,
        NUM_UNIFORMS
    };
    GLint uniforms[NUM_UNIFORMS];
    
    // Attribute index.
    enum
    {
        ATTRIB_VERTEX,
        ATTRIB_NORMAL,
        NUM_ATTRIBUTES
    };
}

/**
 * Initializor of the game shader where the vertex and fragment shader of
 * the game engine are loaded
 *
 */
- (id)init {
    self = [super init: @"entity_vertex_shader" : @"entity_fragment_shader"];
    return self;
}

/**
 * Called to bind the attributes to the program shader
 */
- (void) bindAttributes {
    [super bindAttribute : LOCATION_ATTR_ID: "position"];
    [super bindAttribute : TEX_COORDINATE_ATTR_ID: "textureCoords"];
    [super bindAttribute : NORMAL_VECTOR_ATTR_ID: "normal"];
}

/**
 * Called to ensure that all the shader managers get their uniform locations
 */
- (void) getAllUniformLocations {
    @autoreleasepool {
        NSDictionary * uniformsDic = [NSDictionary dictionaryWithObjectsAndKeys:
                                      @"projectionMatrix", [NSNumber numberWithInt:PROJECTION_MATRIX],
                                      @"viewMatrix", [NSNumber numberWithInt: VIEW_MATRIX],
                                      @"transformationMatrix", [NSNumber numberWithInt: TRANSFORMATION_MATRIX],
                                      @"lightPosition", [NSNumber numberWithInt:LIGHT_POSITION],
                                      @"lightColor", [NSNumber numberWithInt:LIGHT_COLOR],
                                      @"shineDamper", [NSNumber numberWithInt:SHINE_DAMPER],
                                      @"reflectivity", [NSNumber numberWithInt:REFLECTIVITY],
                                      @"skyColor", [NSNumber numberWithInt:SKY_COLOR],
                                      @"normalsPointingUp", [NSNumber numberWithInt:NORMALS_POINTING_UP],
                                      nil];
        for(id key in uniformsDic) {
            int iKey = [key intValue];
            const char *shaderVariable = [(uniformsDic[key]) UTF8String];
            uniforms[iKey] = [super getUniformLocation: shaderVariable];
            if(uniforms[iKey] < 0) {
                NSLog(@"Problems getting %@'s location in the shader", uniformsDic[key]);
            }
        }
    }
}


/**
 * Load the projection matrix
 *
 * @param matrix
 *            the matrix to be loaded
 */
- (void) loadProjectionMatrix : (GLTransformation*) matrix {
    [super loadMatrix:  uniforms[PROJECTION_MATRIX]: matrix];
}

/**
 * Load the view matrix
 *
 * @param matrix
 *            the matrix to be loaded
 */
- (void) loadViewMatrix : (GLTransformation*) matrix {
    [super loadMatrix:  uniforms[VIEW_MATRIX]: matrix];
}

/**
 * Load the transformation matrix
 *
 * @param matrix
 *            the matrix to be loaded
 */
- (void) loadTransformationMatrix : (GLTransformation*) matrix {
    [super loadMatrix:  uniforms[TRANSFORMATION_MATRIX]: matrix];
}

/**
 * Put passes the information of the light to the
 * Shader program
 *
 * @param light the light to load in the shader program
 */
- (void) loadLight : (Light*) light {
    [super loadVector: uniforms[LIGHT_POSITION] : [light position]];
    [super loadVector: uniforms[LIGHT_COLOR] : [light color]];
}

/**
 * Load the values of the specular light in the fragment shader
 *
 * @param damper		The damper of the specular lighting
 * @param reflectivity	The reflectivity of the material
 */
- (void) loadShineVariables : (float) damper : (float) reflectivity {
    [super loadFloat : uniforms[SHINE_DAMPER] : damper];
    [super loadFloat : uniforms[REFLECTIVITY] : reflectivity];
}

/**
 * Set in the shader if the normals should all of them point up
 *
 * @param normalsPointingUp Flag that indicates if all the normals of the entity are poiting up or not
 */
- (void)  loadNormalsPointingUp : (Boolean) normalsPointingUp {
    [super loadBoolean : uniforms[NORMALS_POINTING_UP] :  normalsPointingUp];
}

/**
 * Load the color of the sky in order to simulate fog
 *
 * @param skyColor
 * 			Color of the sky
 */
- (void)  loadSkyColor : (Vector3f*) skyColor {
    [super loadVector: uniforms[SKY_COLOR] : skyColor];
}

@end

File: GameEngine_IOS/GameEngine/Shaders/Entities/EntityShaderManager.swift
import Foundation

open class EntityShaderManager : ShaderManager {
    /**
     * Initializor of the game shader where the vertex and fragment shader of
     * the game engine are loaded
     *
     */
    public init() {
        super.init(vertexFile: "entity_vertex_shader" , fragmentFile: "entity_fragment_shader", numberOfUniforms: TEntityUniform.numOfEntityLocations.rawValue)
    }
    
    /**
     * Called to bind the attributes to the program shader
     */
    override internal func getAttributes() -> Dictionary<Int, String>! {
        let attributesDic : Dictionary<Int, String> = [
            TEntityAttribute.position.rawValue : "\(TEntityAttribute.position)",
            TEntityAttribute.textureCoords.rawValue : "\(TEntityAttribute.textureCoords)",
            TEntityAttribute.normal.rawValue : "\(TEntityAttribute.normal)"
        ];
        return attributesDic;
    }
    
    /**
     * Called to provide the uniform locations that are going to get bound to the shader
     */
    override internal func getUniformLocations() -> Dictionary<NSInteger, String>! {
        let uniformsDic : Dictionary<NSInteger, String> = [
            TEntityUniform.projectionMatrix.rawValue : "\(TEntityUniform.projectionMatrix)",
            TEntityUniform.viewMatrix.rawValue : "\(TEntityUniform.viewMatrix)",
            TEntityUniform.transformationMatrix.rawValue : "\(TEntityUniform.transformationMatrix)",
            TEntityUniform.lightPosition.rawValue : "\(TEntityUniform.lightPosition)",
            TEntityUniform.lightColor.rawValue : "\(TEntityUniform.lightColor)",
            TEntityUniform.shineDamper.rawValue : "\(TEntityUniform.shineDamper)",
            TEntityUniform.reflectivity.rawValue : "\(TEntityUniform.reflectivity)",
            TEntityUniform.skyColor.rawValue : "\(TEntityUniform.skyColor)",
            TEntityUniform.normalsPointingUp.rawValue : "\(TEntityUniform.normalsPointingUp)",
            TEntityUniform.textureWeight.rawValue : "\(TEntityUniform.textureWeight)",
            TEntityUniform.diffuseColor.rawValue : "\(TEntityUniform.diffuseColor)"
        ]
        
        return uniformsDic;
    }
    
    
    /**
     * Load the projection matrix
     *
     * @param matrix
     *            the matrix to be loaded
     */
    open func loadProjectionMatrix (_ matrix : GLTransformation) {
        super.loadMatrix(uniforms[TEntityUniform.projectionMatrix.rawValue], matrix: matrix);
    }
    
    /**
     * Load the view matrix
     *
     * @param matrix
     *            the matrix to be loaded
     */
    open func loadViewMatrix (_ matrix : GLTransformation) {
        super.loadMatrix(uniforms[TEntityUniform.viewMatrix.rawValue], matrix: matrix);
    }
    
    /**
     * Load the transformation matrix
     *
     * @param matrix
     *            the matrix to be loaded
     */
    open func loadTransformationMatrix(_ matrix : GLTransformation) {
        super.loadMatrix(uniforms[TEntityUniform.transformationMatrix.rawValue], matrix: matrix);
    }
    
    /**
     * Put passes the information of the light to the
     * Shader program
     *
     * @param light the light to load in the shader program
     */
    open func loadLight(_ light : Light) {
        super.loadVector(uniforms[TEntityUniform.lightPosition.rawValue], vector: light.position);
        super.loadColorRGB(uniforms[TEntityUniform.lightColor.rawValue], color: light.color);
    }
    
    /**
     * Load the values of the specular light in the fragment shader
     *
     * @param damper		The damper of the specular lighting
     * @param reflectivity	The reflectivity of the material
     */
    open func loadShineVariables(_ damper : Float, reflectivity : Float) {
        super.loadFloat(uniforms[TEntityUniform.shineDamper.rawValue], value: damper);
        super.loadFloat(uniforms[TEntityUniform.reflectivity.rawValue], value : reflectivity);
    }
    
    /**
     * Set in the shader if the normals should all of them point up
     *
     * @param normalsPointingUp Flag that indicates if all the normals of the entity are poiting up or not
     */
    open func loadNormalsPointingUp(_ normalsPointingUp : Bool) {
        super.loadBoolean( uniforms[TEntityUniform.normalsPointingUp.rawValue], value :  normalsPointingUp);
    }
    
    /**
     * Load the color of the sky in order to simulate fog
     *
     * @param skyColor
     * 			Color of the sky
     */
    open func  loadSkyColor(_ skyColor : ColorRGBA) {
        super.loadColorRGBA(uniforms[TEntityUniform.skyColor.rawValue], color: skyColor)
    }
    
    /**
     * Load the texture weight of the material
     *
     * @param textureWeight
     *            texture weight of the material
     */
    open func loadTextureWeight(_ textureWeight : Float) {
        super.loadFloat(uniforms[TEntityUniform.textureWeight.rawValue], value: textureWeight);
    }
    
    /**
     * Load the diffuse color of the material
     *
     * @param diffuseColor
     *            diffuse color of the material
     */
    open func loadDiffuseColor(_ diffuseColor : ColorRGBA) {
        super.loadColorRGBA(uniforms[TEntityUniform.diffuseColor.rawValue], color : diffuseColor);
    }
}
